service/client_state.cc

```diff
@@ -270,3 +271,22 @@ future<> service::client_state::ensure_exists(const auth::resource& r) const {
         return make_ready_future<>();
     });
 }
+
+future<> service::client_state::maybe_update_per_service_level_params() {
+    if (_sl_controller && _user && _user->name) {
+        const auto& role_manager = _auth_service->underlying_role_manager();
+        auto role_set = co_await role_manager.query_granted(_user->name.value(), auth::recursive_role_query::yes);
+        auto slo_opt = co_await _sl_controller->find_service_level(role_set);
+        if (!slo_opt) {
+            co_return;
+        }
+        const qos::service_level_options& slo = *slo_opt;
```

Comment:
I see why we need _sl_controller now. Not too happy about it, but it can remain.

Maybe _sl_controller should keep track of its service levels and propagate changes to them, instead of the client states polling for changes.